### PR TITLE
style: add EditorConfig and format some files

### PR DESCRIPTION
**Pull request by flyingthingsintothings**
Add an EditorConfig to make sure editors know how to correctly edit and save files. Also strip trailing whitespace in some files that don't need it, as configured in the EditorConfig file.

### CODE REVIEW

**Comment by flyingthingsintothings** [comment]
I checked files with grep and find to see what extensions there were, which files used what conventions and tried to add all the conventions in the EditorConfig file to enforce a more consistent style in the future. I also limited `trim_trailing_whitespace` to files that would normally never need trailing whitespace.

Having this file is especially handy when this code is used as a submodule where another EditorConfig file might set different settings. The editor always uses the EditorConfig that is closest to the edited file, which would be the one added here.

**Comment by hamishwillee** [comment]
@flyingthingsintothings Thank you so much - this is great. In particular, should fix some of those spurious formatting errors when working with new message definitions,

@julianoes I'm approving. But good to have a second pair of eyes so would appreciate you could do a check too.

**Review by hamishwillee — APPROVED** [review]

File: .editorconfig [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+[*.json]
+tab_width = 2
+trim_trailing_whitespace = true
+
+[*.{yml,yaml}]
+tab_width = 2
+
+[*.{c,cpp,cc,h,hpp}]
+indent_style = tab
+tab_width = 8
```
Comment (julianoes) [inline .editorconfig:39]: I'd probably do 4 here.

**Review by julianoes — COMMENTED** [review]
Fine with me.

**Comment by julianoes** [comment]
@hamishwillee would be nice if your editor wouldn't leave trailing whitespace on save.

**Comment by flyingthingsintothings** [comment]
I added some rules to remove trailing whitespace on save but didn't do it for all files as I know it can be meaningful in some languages.

**Review by julianoes — APPROVED** [review]

**Comment by hamishwillee** [comment]
> @hamishwillee would be nice if your editor wouldn't leave trailing whitespace on save.

It certainly would :-)

Thanks @flyingthingsintothings - merging.

**Pull request merged**